sklearn/manifold/_utils.pyx

```diff
@@ -12,8 +12,8 @@ cdef extern from "numpy/npy_math.h":
 cdef float EPSILON_DBL = 1e-8
 cdef float PERPLEXITY_TOLERANCE = 1e-5
 
-cpdef cnp.ndarray[cnp.float32_t, ndim=2] _binary_search_perplexity(
-        cnp.ndarray[cnp.float32_t, ndim=2] sqdistances,
+def _binary_search_perplexity(
+        cnp.float32_t[:, :] sqdistances,
         float desired_perplexity,
         int verbose):
     """Binary search for sigmas of conditional Gaussians.
```

Comment:
Can you as well update the docstring:

```diff
@@ -23,7 +20,7 @@ cpdef cnp.ndarray[cnp.float32_t, ndim=2] _binary_search_perplexity(
 
     Parameters
     ----------
-    sqdistances : array-like, shape (n_samples, n_neighbors)
+    sqdistances : ndarray of shape (n_samples, n_neighbors), dtype=np.float32
         Distances between training samples and their k nearest neighbors.
         When using the exact method, this is a square (n_samples, n_samples)
         distance matrix. The TSNE default metric is "euclidean" which is
@@ -37,7 +34,7 @@ cpdef cnp.ndarray[cnp.float32_t, ndim=2] _binary_search_perplexity(
 
     Returns
     -------
-    P : array, shape (n_samples, n_samples)
+    P : ndarray of shape (n_samples, n_samples), dtype=np.float32
         Probabilities of conditional Gaussian distributions p_i|j.
     """
```

Reply:
It seems that this use `np.float64` for `P`.